File: +fs/+csr/buildCorners.m
function FS = buildCorners(FS)
%BUILDCORNERS  Construct the CSR-flat node→corner layout.
%
%   FS = fs.csr.buildCorners(FS)
%
%   Populates:
%     FS.csr.nCorners      scalar — total corners = sum(esurn2-diff)
%     FS.csr.cornerNode    [nCorners x 1] — which node owns each corner
%     FS.csr.cornerElem    [nCorners x 1] — which element each corner sits in
%     FS.csr.nodePtr       [nNodes+1 x 1] — alias of esurn2 (row-pointer)
%     FS.csr.nodeNec       [nNodes x 1]   — corners per node
%     FS.csr.maxNec        scalar — max corners on any single node
%
%   This layout is the KEY vectorization enabler: with it, every LPEW2 corner
%   computation becomes a flat vector op (see study/artifacts/data-structures.md).

    if ~isfield(FS.mesh, 'esurn1') || ~isfield(FS.mesh, 'esurn2')
        error('fs.csr.buildCorners: FS.mesh.esurn1/esurn2 missing');
    end

    esurn1 = FS.mesh.esurn1(:);
    esurn2 = FS.mesh.esurn2(:);
    nNodes = FS.mesh.nNodes;

    FS.csr.nodePtr    = esurn2;
    FS.csr.nCorners   = numel(esurn1);
    FS.csr.cornerElem = esurn1;
    FS.csr.nodeNec    = diff(esurn2);
    if numel(FS.csr.nodeNec) ~= nNodes
        error('fs.csr.buildCorners: diff(esurn2) has %d entries, expected nNodes=%d', ...
              numel(FS.csr.nodeNec), nNodes);
    end
    FS.csr.cornerNode = repelem((1:nNodes).', FS.csr.nodeNec);
    FS.csr.maxNec     = max(FS.csr.nodeNec);
end

File: fs/csr/buildCorners.py
"""CSR corner layout translated from ``+fs/+csr/buildCorners.m``."""

from typing import Any, Dict

import numpy as np


def buildCorners(FS: Dict[str, Any]) -> Dict[str, Any]:
    """Populate the flat node-to-corner layout in ``FS['csr']``."""
    mesh = FS["mesh"]
    if "esurn1" not in mesh or "esurn2" not in mesh:
        raise ValueError("FS.mesh.esurn1 and FS.mesh.esurn2 are required")

    esurn1 = np.asarray(mesh["esurn1"], dtype=int).reshape(-1)
    esurn2 = np.asarray(mesh["esurn2"], dtype=int).reshape(-1)
    n_nodes = int(mesh["nNodes"])
    node_nec = np.diff(esurn2)

    if node_nec.size != n_nodes:
        raise ValueError(
            f"diff(esurn2) has {node_nec.size} entries, expected {n_nodes}"
        )
    if esurn2[0] != 0 or esurn2[-1] != esurn1.size or np.any(node_nec < 0):
        raise ValueError("esurn2 is not a valid CSR row pointer")

    csr = FS.setdefault("csr", {})
    csr["nodePtr"] = esurn2
    csr["nCorners"] = esurn1.size
    csr["cornerElem"] = esurn1
    csr["nodeNec"] = node_nec
    csr["cornerNode"] = np.repeat(np.arange(n_nodes), node_nec)
    csr["maxNec"] = int(node_nec.max()) if node_nec.size else 0
    return FS
